cpp/dolfinx/common/array_2d.h

```diff
@@ -0,0 +1,187 @@
+           const Allocator& alloc = Allocator())
+      : _rows(rows), _cols(columns)
+  {
+    if (_rows * _cols != list.size())
+      throw std::runtime_error("Dimension mismatch");
+    _storage = std::vector<T, Allocator>(list, alloc);
+  }
+
+  /// Returns a reference to the element at specified location (i, j).
+  /// No bounds checking is performed.
+  reference operator()(size_type i, size_type j)
+  {
+    return _storage[i * _cols + j];
+  }
+
+  /// Returns a const reference to the element at specified location (i, j).
+  /// No bounds checking is performed.
+  const_reference operator()(size_type i, size_type j) const
+  {
+    return _storage[i * _cols + j];
+  }
+
+  /// Returns a reference to the row at specified location (i).
+  tcb::span<value_type> row(int i) const
+  {
+    size_type offset = i * _cols;
+    return tcb::span<value_type>(pointer(&_storage[0] + offset), _cols);
+  }
+
+  /// Returns a pointer to the first element of the array.
+  value_type* data() noexcept { return _storage.data(); }
+
+  /// Returns a const pointer to the first element of the array.
+  const value_type* data() const noexcept { return _storage.data(); };
+
+  /// Returns an iterator to the first element of the array.
+  iterator begin() noexcept { return _storage.begin(); }
+
+  /// Returns an iterator to the first element of the array.
+  const_iterator begin() const noexcept { return _storage.begin(); }
+
+  /// Returns an iterator to the element following the last element.
+  iterator end() noexcept { return _storage.end(); }
+
+  /// Returns an iterator to the element following the last element.
+  const_iterator end() const noexcept { return _storage.end(); }
+
+  /// Returns a const iterator to the first element of the array.
+  const_iterator cbegin() noexcept { return _storage.cbegin(); }
+
+  /// Returns a const iterator to the element following the last element.
+  const_iterator cend() noexcept { return _storage.cend(); }
+
+  /// Returns the number of elements in the array (rows * cols).
+  size_type size() const noexcept { return _storage.size(); }
+
+  /// Returns the number of cols in the two-dimensional array.
+  size_type cols() const noexcept { return _cols; }
+
+  /// Returns the number of rows in the two-dimensional array.
+  size_type rows() const noexcept { return _rows; }
+
+  /// Returns the array dimensions {cols, rows}.
+  std::pair<size_type, size_type> shape() const noexcept
+  {
+    return {_rows, _cols};
+  }
+
+  /// Changes the number of elements stored.
+  void resize(size_type rows, size_type cols, value_type val = value_type())
+  {
+    // TODO: check rows and cols
+    _rows = rows;
+    _cols = cols;
+    _storage.resize(rows * cols, val);
+  }
+
+  /// Copies a block of data from other container.
+  template <class Container,
+            typename
+            = typename std::enable_if<is_2d_container<Container>::value>::type>
+  void copy(const Container& other)
```

Comment:
Not sure about the name - I associated `copy` with a new object. Maybe `assign` is better?

Reply:
Changed function name to `assign`.